src/main/java/com/training/tinyurl/controller/TinyUrlControllerImpl.java

```diff
@@ -61,8 +67,56 @@ public ResponseEntity<String> logoutUser(@AuthenticationPrincipal AppUserDetails
     @PutMapping("upgradePlan")
     @PreAuthorize("hasAuthority('BASIC')")
     @Override
-    public ResponseEntity<String> upgradePlan(){
-        tinyUrlService.upgradePlan();
+    public ResponseEntity<String> upgradePlan(@AuthenticationPrincipal AppUserDetails user){
+        tinyUrlService.upgradePlan(user);
+        log.info("Plan upgrade for "+user.getEmail());
         return new ResponseEntity<>("upgrade successful please login again",HttpStatus.OK);
     }
+
+    @PostMapping("getTinyUrl")
+    @Override
+    public ResponseEntity<String> getTinyUrl(@RequestBody @Valid TinyUrlDto dto, BindingResult result,
+                                             @AuthenticationPrincipal AppUserDetails user)
+                                throws ValidationException, NoSuchAlgorithmException {
+        Validator.validate(result);
+        String tinyurl=tinyUrlService.genTinyUrl(dto.getLongUrl(),user);
+        log.info("Generated tinyurl : " +tinyurl);
+        return new ResponseEntity<>(tinyurl,HttpStatus.OK);
+    }
+
+    @GetMapping("fetch")
```

Comment:
Your full api name is /api/v1/user/fetch and /api/v1/user/delete / update and /redirect but actually you are not deleting, updating, fetching user.

That is why you need to rely on API name as nouns and not verbs.

Create Tiny URL - POST /api/v1/user/{userId}/url
Get Long URL from Tiny Url - GET /api/v1/user/{userId}/url/{tinyUrl}
Update Tiny URL - PUT /api/v1/user/{userId}/url/{tinyUrl}
Delete Tiny URL - DELETE /api/v1/user/{userId}/url/{tinyUrl}